clients/ruby/zawgyidetector/lib/zawgyidetector/zawgyi_unicode_markov_model.rb

```diff
@@ -0,0 +1,182 @@
+# Copyright 2018 Aung Kyaw Phyo
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A Markov model to predict whether a string is more likely Unicode or Zawgyi.
+
+# Internally, this class maintains two Markov chains, one representing Unicode and the other
+# representing Zawgyi. An input string is evaluated against both chains, and the chain that returns
+# the higher probability becomes the prediction.
+
+# A string is evaluated as a sequence of transitions between states, including transitions to
+# the edges of the string. For example, the string "ABC" contains 4 transitions: NULL to A, A to B,
+# B to C, and C to NULL.
+
+# For the purposes of Unicode/Zawgyi detection, all characters are treated as the NULL state
+# except for characters in the Myanmar script or characters in the Unicode whitespace range U+2000
+# through U+200B.
+class ZawgyiUnicodeMarkovModel
+
+    # Magic number used to identify this object in byte streams. (Reads in ASCII as "UZMODEL ")
+    BINARY_TAG = "555a4d4f44454c20"
+  
+    # Current serial format version number, used in association with the magic number.
+    BINARY_VERSION = 1
+  
+    # Standard Myanmar code point range before digits
+    STD_CP0 = 4096 # "\u1000"
```

Comment:
Why not write these as hexadecimal literals?

0x1000
0x103F
0x104A
...

Reply:
Thank you for the review and feedbacks.

Sure. I will write these code points in hexadecimal literals.